File: Python/LeetCodeTest/lc055_JumpGame.py
from typing import List

class Solution:
    def canJump(self, nums: List[int]) -> bool: # 深度搜索
        n = len(nums)
        if n < 2:
            return True
        cur_max = 0
        next_max = 0
        i = 0
        while cur_max - i + 1 > 0:
            while i <= cur_max: # 能触及的最深位置
                next_max = max(next_max, nums[i] + i) # 下一个的最深位置
                if next_max >= n - 1: # 超过终点就返回
                    return True
                i += 1
            cur_max = next_max
        return False

if __name__ == "__main__":
    solution = Solution()
    result = solution.canJump([2, 3, 1, 1, 4])
    print(result)

File: Java/src/com/liu/LeetCode/CodeCaprice/Greedy/LC055_JumpGame.java
package com.liu.LeetCode.CodeCaprice.Greedy;

public class LC055_JumpGame {

    // Greedy
    // Beats 79.75%
    private static boolean canJumpMy1(int[] nums) {
        int length = nums.length;
        int maxReach = 0;
        for (int i = 0; i < length; i++) {
            if (i > maxReach) {
                // 到不了当前的i位置
                return false;
            }
            maxReach = Math.max(maxReach, i + nums[i]);
        }
        // 已经可以到最后一个位置了
        return maxReach >= length - 1;
    }

    public static void main(String[] args) {
        int[][] pricesList = new int[][]{
                {2, 3, 1, 1, 4},
                {3, 2, 1, 0, 4},
        };
        boolean result;
        long startTime;
        long endTime;

        for (int i = 0; i < pricesList.length; i++) {
            startTime = System.currentTimeMillis();
            result = canJumpMy1(pricesList[i]);
            endTime = System.currentTimeMillis();
            System.out.println("canJump My1 result" + i + " " + result + " during time " + (endTime - startTime));
        }
    }
}
